### PR TITLE
refactor: WriteAcknowledgement API

### PR DESCRIPTION
**Pull request by seantking**
## Description
Updating WriteAcknowledgement API to take `exported.Acknowledgement` type instead of `[ ] bytes` 

closes: #854 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/core/04-channel/keeper/packet.go [hunk 1]
```diff
@@ -342,14 +342,12 @@ func (k Keeper) WriteAcknowledgement(
 		return types.ErrAcknowledgementExists
 	}
 
-	if len(acknowledgement) == 0 {
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/packet.go:345]: How come the length check is removed? Should this not be kept?

**Review by colin-axner — COMMENTED** [review]
Nice!

The ack length check needs to be added back. A changelog entry is needed and an entry into the migration docs for IBC app developers. We could backport to v3, but it isn't needed for transfer/ica so maybe best to leave for v4 instead of introducing changes for IBC app developers to implement (so make a new migration doc, you can copy the template)

**Comment by seantking** [comment]
> Nice!
> 
> The ack length check needs to be added back. A changelog entry is needed and an entry into the migration docs for IBC app developers. We could backport to v3, but it isn't needed for transfer/ica so maybe best to leave for v4 instead of introducing changes for IBC app developers to implement (so make a new migration doc, you can copy the template)

Just added back the ack length check (I wrongly assumed an Ack wouldn't have empty bytes). 

I'll do the changelog and the docs update. Good catch, thanks :+1:

File: modules/core/04-channel/keeper/packet.go (hunk 1 above)
Comment (seantking) [reply modules/core/04-channel/keeper/packet.go:345]: Yeah thanks, good catch. I was working under the assumption that an Acknowledgement wouldn't have empty ack bytes

**Review by damiannolan — APPROVED** [review]

**Comment by seantking** [comment]
@AdityaSripal @colin-axner So I've ran into an annoying issue with this. 

We aren't able to leave in the check to check for empty byte array i.e. `len(ackbytes) == 0` as calling `ack.Acknowledgement()` will always have len > 0 due to how we're serializing [into JSON](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/types/acknowledgement.go#L61) (I guess there is extra padding because of the JSON). 

Am I correct in saying the main reason we want to do this update to the API is for [code cleanliness](https://github.com/cosmos/ibc-go/pull/749/files#diff-60916badc8c21c745d39164a632c72516e493179a3a7bb3707c3aa682df6e019R26 ). I guess we have to pack and then unpack the ack bytes if we don't have this change, which is a bit ugly?

File: modules/core/04-channel/keeper/packet.go [hunk 2]
```diff
@@ -342,14 +342,15 @@ func (k Keeper) WriteAcknowledgement(
...
+	bz := acknowledgement.Acknowledgement()
+	if len(bz) == 0 {
 		return sdkerrors.Wrap(types.ErrInvalidAcknowledgement, "acknowledgement cannot be empty")
```
Comment (AdityaSripal) [inline modules/core/04-channel/keeper/packet.go:347]: Just change this check to `if acknowledgement == nil`

Comment (seantking) [reply modules/core/04-channel/keeper/packet.go:347]: So this is what I had originally. 

Can we go with this @damiannolan @colin-axner?

File: testing/mock/ack.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package mock
+
+// MockEmptyAcknowledgement implements the exported.Acknowledgement interface and always returns an empty byte string as Response
+type MockEmptyAcknowledgement struct {
```
Comment (seantking) [inline testing/mock/ack.go:4]: @colin-axner Open to suggestions with this. I don't want to add technical debt unnecessarily.

**Comment by seantking** [comment]
@AdityaSripal @colin-axner I added a `MockEmptyAcknowledgement` type which implements the `exported.Acknowledgement` interface. We can use this in the tests to ensure we test the check for returning an error if ack bytes length == 0. 

Open to ideas for alternative workarounds.

File: testing/mock/ack.go (hunk 3 above)
Comment (colin-axner) [reply testing/mock/ack.go:4]: No this is great. None of the mock code has added any technical debt so far. I think it'll be useful to keep around

**Review by colin-axner — APPROVED** [review]
Excellent!

File: docs/migrations/v3-to-v4.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+There are four sections based on the four potential user groups of this document:
+- Chains
+- IBC Apps
+- Relayers
+- IBC Light Clients
+
+**Note:** ibc-go supports golang semantic versioning and therefore all imports must be updated to bump the version number on major releases.
+```go
+github.com/cosmos/ibc-go/v2 -> github.com/cosmos/ibc-go/v3
+```
```
Comment (AdityaSripal) [inline docs/migrations/v3-to-v4.md:15]: Shouldn't this be v3->v4?

File: modules/core/04-channel/keeper/packet.go [hunk 5]
```diff
@@ -342,14 +342,15 @@ func (k Keeper) WriteAcknowledgement(
 		return types.ErrAcknowledgementExists
 	}
 
-	if len(acknowledgement) == 0 {
+	bz := acknowledgement.Acknowledgement()
+	if len(bz) == 0 {
```
Comment (AdityaSripal) [inline modules/core/04-channel/keeper/packet.go:346]:
```suggestion
	if acknowledgement == nil || len(acknowledgement.Acknowledgement()) == 0 {
```

File: modules/core/04-channel/keeper/packet.go [hunk 6]
```diff
@@ -342,14 +342,15 @@ func (k Keeper) WriteAcknowledgement(
 		return types.ErrAcknowledgementExists
 	}
 
-	if len(acknowledgement) == 0 {
+	bz := acknowledgement.Acknowledgement()
```
Comment (AdityaSripal) [inline modules/core/04-channel/keeper/packet.go:345]: You should still check if acknowledgement is nil first or it will panic

**Review by AdityaSripal — CHANGES_REQUESTED** [review]
Still need a nil check. Everything else looks great!

File: docs/migrations/v3-to-v4.md (hunk 4 above)
Comment (seantking) [reply docs/migrations/v3-to-v4.md]: yep

File: modules/core/04-channel/keeper/packet.go (hunk 6 above)
Comment (seantking) [reply modules/core/04-channel/keeper/packet.go:349]: Good call, thank you

File: modules/core/04-channel/keeper/packet.go (hunk 5 above)
Comment (seantking) [reply modules/core/04-channel/keeper/packet.go:350]: Added the check as a separate if statement. Just my personal preference.

**Review by AdityaSripal — APPROVED** [review]

**Comment by colin-axner** [comment]
I think we should backport this to v3. Does anyone disagree?

**Pull request merged**